fix(compiler): check more cases for pipe usage inside host bindings

crisbeto opened the pull request:
Builds on top of #34655 to support more cases that could be using a pipe inside host bindings (e.g. ternary expressions or function calls).

Fixes #37610.

JoostK commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -3090,6 +3090,142 @@ runInEachFileSystem(os => {
              .toContain('Host binding expression cannot contain pipes');
        });
 
+    it('should throw if a pipe is used inside a conditional in a host binding', () => {
```
I think these detailed tests would be more appropriate as simple unit tests here:

https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/compiler/test/expression_parser/parser_spec.ts#L728-L732

JoostK commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -1093,11 +1093,46 @@ class SimpleExpressionChecker implements AstVisitor {
 
   visitPrefixNot(ast: PrefixNot, context: any) {
-    ast.expression.visit(this);
+    ast.expression.visit(this, context);
```
I think it's best for this class to use `RecursiveAstVisitor` as base class. That would require that `SimpleExpressionChecker` becomes an interface, but that should be a quick change to make.

JoostK commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -1093,11 +1093,46 @@ class SimpleExpressionChecker implements AstVisitor {
+    ast.falseExp.visit(this, context);
+  }
+
+  visitFunctionCall(ast: FunctionCall, context: any) {
+    this.visitAll(ast.args, context);
+  }
+
+  visitMethodCall(ast: MethodCall, context: any) {
+    this.visitAll(ast.args, context);
+  }
+
+  visitSafeMethodCall(ast: SafeMethodCall, context: any) {
+    this.visitAll(ast.args, context);
+  }
```
These are missing the visits of `target`/`receiver`. One of the reasons why I'd prefer we switch this over to `RecursiveAstVisitor` :-)

crisbeto committed:
fix(compiler): check more cases for pipe usage inside host bindings

Builds on top of #34655 to support more cases that could be using a pipe inside host bindings (e.g. ternary expressions or function calls).

Fixes #37610.

crisbeto commented:
Thanks for the feedback @JoostK, the recursive visitor definitely cleaned things up 👍. Can you take another look?

JoostK reviewed (approved):
LGTM
